lib/AST/StmtManual.cpp

```diff
@@ -65,55 +65,166 @@ std::optional<Macro> Stmt::HighestContainingSubstitution(void) const noexcept {
   return std::nullopt;
 }
 
-// Returns true if we can follow the given token's derived locations to a reach
-// a token that has a macro location whose parent is the specified macro.
-inline bool IsTokenDerivedFromMacro(const Token &token, const Macro &macro) {
-  for (auto derived_tok = std::optional(token); derived_tok;
-       derived_tok = derived_tok->DerivedLocation()) {
-    auto macro_tok = derived_tok->MacroLocation();
-    if (!macro_tok) {
-      continue;
-    }
-
-    auto containing_macro = macro_tok->Parent();
-    if (macro == containing_macro.value()) {
-      return true;
-    }
-  }
-  return false;
-}
-
 // Returns the lowest macro argument that contains this Stmt, if any.
 std::optional<MacroArgument>
 Stmt::LowestContainingMacroArgument(void) const noexcept {
   // Algorithm:
   // 1. Find the lowest macro argument that the first token in this Stmt was
   //    expanded from.
-  // 2. Check if every other token in the Stmt was also expanded from this
-  //    argument at some point.
+  // 2. Check if the last token in the Stmt was also expanded from this argument
+  //    at some point.
   // 3. If so, then return the argument that the first token was expanded from.
   // 4. Otherwise, find the next-lowest macro argument that the first token was
-  //    expanded from and repeat steps 2 and 3.
-  //    If we reach a point where the first token is no longer expanded from a
-  //    macro argument, then return std::nullopt.
+  //    expanded from and repeat steps 2 and 3. If we reach a point where the
+  //    first token is no longer expanded from a macro argument, then return
+  //    std::nullopt.
+  
+  const auto last_tok = *(Tokens().end() - 1);
```

Comment:
It may be that `Tokens()` is empty. I would double check this. I would assign the return value of `Tokens()` to a variable, then check the size, return nullopt if empty, then use the var for getting at the last one too.

Reply:
I think you added `.Front()` and `.Back()` to make that easier? You could use the optional return value of `.Front()` to tell you that you have no tokens, then assume `.Back()` is non-`nullopt`.

Reply:
Ah I never realized that a Stmt could have an empty token list - that's what I get for assuming :sweat_smile: 

And yes I defined `Front()` and `Back()` for facilitating these sorts of accesses, but only for `MacroRange`. I'll define the same methods for `TokenRange` and use them here.